### PR TITLE
Fix T5 and BART for TF

### PR DESCRIPTION
**Pull request by jplu**
# What does this PR do?

This PR fix the TensorFlow implementation of T5 and BART to make them graph compilation+execution compliant and then be able to create a savedmodel for each them.

The slow tests `test_saved_model_with_hidden_states_output` and `test_saved_model_with_attentions_output` are now passing for both models.

### CODE REVIEW

File: tests/test_modeling_tf_common.py
Comment (sgugger) [inline tests/test_modeling_tf_common.py:339]: Any reason to change a bool to a tensor? We can just keep it as a boolean, no?

**Review by sgugger — APPROVED** [review]
Way cleaner! Thanks a lot for doing this!

File: tests/test_modeling_tf_common.py
Comment (jplu) [reply tests/test_modeling_tf_common.py:339]: This is because now, the `use_cache` parameter returned by the `prepare_inputs` method in the T5 test file is now a primitive boolean and not anymore a tensor. Thanks to the changes the value is not forced to be a tensor anymore :)

Comment (sgugger) [reply tests/test_modeling_tf_common.py:339]: Yes I understand that. PyTorch accepts regular booleans in its models, that's why I'm confused why we convert it to a PyTorch tensor here.

Comment (jplu) [reply tests/test_modeling_tf_common.py:339]: This is because of the `.numpy()` to convert TF tensors to numpy array. Simple booleans don't have this attribute.

Comment (sgugger) [reply tests/test_modeling_tf_common.py:339]: My question is why not just pass `pt_inputs_dict[name] = key`

Comment (jplu) [reply tests/test_modeling_tf_common.py]: Ah ok!! Simply because I didn't know 😄 Just pushed the update^^

File: src/transformers/models/bart/modeling_tf_bart.py
Comment (LysandreJik) [inline src/transformers/models/bart/modeling_tf_bart.py:575]: Does removing this mean the `TFBartDecoder` will not be able to handle the output attentions/hidden states/cache/return dict parameters on its own, but will instead rely on being called by the class `TFBart{Model, ForConditionalGeneration}`?

If @patrickvonplaten is doing the same refactor for TF BART as he did for the PT version, then one of the enhancements it offers is being able to use the `(TF)BartDecoder` as a standalone model. Will this change prevent this?

File: src/transformers/models/t5/modeling_tf_t5.py
Comment (LysandreJik) [inline src/transformers/models/t5/modeling_tf_t5.py:1164]: Infinitely cleaner

**Review by LysandreJik — APPROVED** [review]
Overall way cleaner. Happy to see the `cast_bool_to_primitive` go!
Thanks for working on this.

File: src/transformers/models/bart/modeling_tf_bart.py
Comment (jplu) [reply src/transformers/models/bart/modeling_tf_bart.py:575]: If the `(TF)BartDecoder` aims to be used as a standalone, the `input_processing` must be added. I can add it just in case, it won't arm the current bebavior, just an "over" processing.  Should I do it?

Comment (LysandreJik) [reply src/transformers/models/bart/modeling_tf_bart.py:575]: I think @patrickvonplaten will probably take care of that in his PR, so no problem here. I would still wait for his review before merging!

File: src/transformers/modeling_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:357]: think we can delete `print(k, v)` no?

File: src/transformers/modeling_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:1369]: yeeees :-)

File: src/transformers/models/bart/modeling_tf_bart.py
Comment (patrickvonplaten) [reply src/transformers/models/bart/modeling_tf_bart.py:575]: I haven't looked at having neither TFBart nor TFT5 as "encoder-only" or "decoder-only" model yet because a) there is not TFEncoderDecoder model and b) because of the issues that will be solved in this PR. So I'm 100% fine to delete it for now here

File: src/transformers/models/bart/modeling_tf_bart.py
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_tf_bart.py:1049]: great catch!

File: tests/test_modeling_tf_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_tf_common.py:594]: can we have an `if model.config.is_encoder_decoder` case here instead ? Seems more in line with PT tests and we should test both `decoder_hidden_states` and `encoder_hidden_states` for encoder-decoder

Comment (jplu) [reply tests/test_modeling_tf_common.py]: Better?

**Comment by jplu** [comment]
I should have addressed everybody's comments :)

File: tests/test_modeling_tf_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_tf_common.py:597]: Sorry, I meant something a bit more like this: 

```python
            if model.config.is_encoder_decoder:
                encoder_hidden_states = outputs.encoder_hidden_states
                decoder_hidden_states = outputs.decoder_hidden_states
                
                self.assertEqual(config.output_attentions, False)
                self.assertEqual(len(encoder_hidden_states), expected_num_layers)
                self.assertListEqual(
                    list(encoder_hidden_states[0].shape[-2:]),
                    [self.model_tester.seq_length, self.model_tester.hidden_size],
                )
                self.assertEqual(len(decoder_hidden_states), expected_num_layers)
                self.assertListEqual(
                    list(decoder_hidden_states[0].shape[-2:]),
                    [self.model_tester.seq_length, self.model_tester.hidden_size],
                )
            else:
                hidden_states = outputs.hidden_states
                self.assertEqual(config.output_attentions, False)
                self.assertEqual(len(hidden_states), expected_num_layers)
                self.assertListEqual(
                    list(hidden_states[0].shape[-2:]),
                    [self.model_tester.seq_length, self.model_tester.hidden_size],
                )
```

1) Always good to check for both encoder & decoder for `encoder_decoder` and 2) I prefer to stop using the `self.is_encoder_decoder` flag (the config better defines whether a model is an encoder-decoder - not really the test case IMO)

**Review by patrickvonplaten — APPROVED** [review]
Just have one comment left for the test. After this I think we can merge and I'll rebase my TFBart refactor PR on the new changes here :-)

File: tests/test_modeling_tf_common.py
Comment (jplu) [reply tests/test_modeling_tf_common.py:610]: Ok, just pushed the fix!!!

**Pull request merged**
